<a href="https://colab.research.google.com/github/JulianFernandez1620/AlgorithmsUN2023/blob/main/Lab%204/jufernandezsGroupNLab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB
!./APlusB

2 4
6

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
-5 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
4

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

4

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

4


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

4 5
9


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

4
6 8 22 10
220


### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

90


### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Overwriting input2.txt


In [ ]:
!./max_pairwise_product < input2.txt

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt

90
40000000000


### Test 3  stress tests 20  times  random n in [2, 20]

In [ ]:
%%writefile max_pairwise_product_long_test3.cpp

#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>

using namespace std;

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    srand(time(NULL));

    for (int i = 1; i <= 20; ++i) {
        int n = rand() % 18 + 3;
        vector<int> numbers(n);

        for (int j = 0; j < n; ++j) {
            numbers[j] = rand() % 1000;
        }

        cout << "Caso " << i << ": ";
        for (int j = 0; j < n; ++j) {
            cout << numbers[j] << " ";
        }
        cout << endl;

        int max_product = MaxPairwiseProduct(numbers);
        cout << "El máximo producto de pares es: " << max_product << endl;
    }

    return 0;
}


Overwriting max_pairwise_product_long_test3.cpp


In [ ]:
!g++ max_pairwise_product_long_test3.cpp -o max_pairwise_product_long_test3
!./max_pairwise_product_long_test3

Caso 1: 70 355 455 183 608 635 201 751 623 916 408 896 213 661 119 162 
El máximo producto de pares es: 820736
Caso 2: 242 303 762 804 747 703 669 142 302 263 298 313 411 491 383 118 946 
El máximo producto de pares es: 760584
Caso 3: 727 934 120 478 557 388 238 805 601 899 924 115 153 518 419 916 674 518 971 
El máximo producto de pares es: 906914
Caso 4: 660 273 311 958 586 723 802 
El máximo producto de pares es: 768316
Caso 5: 841 100 888 920 34 8 398 943 748 636 748 350 888 24 817 41 895 236 
El máximo producto de pares es: 867560
Caso 6: 921 106 280 969 766 553 
El máximo producto de pares es: 892449
Caso 7: 77 491 356 231 813 197 331 53 
El máximo producto de pares es: 399183
Caso 8: 718 61 868 
El máximo producto de pares es: 623224
Caso 9: 161 857 410 511 97 786 681 490 
El máximo producto de pares es: 673602
Caso 10: 917 800 955 376 
El máximo producto de pares es: 875735
Caso 11: 924 494 338 205 923 829 561 154 994 111 838 47 581 556 460 801 217 622 10 
El máximo producto de

### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [ ]:
%%writefile Test2Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing Test2Write.cpp


In [ ]:
!g++ Test2Write.cpp -o Test2Write
!./Test2Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

39999800000

real	2m18.847s
user	2m17.594s
sys	0m0.049s


##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [ ]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {
    int n = 10;                                                                                                                            //1
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cout << numbers[i];
    }

    std::cout << "\n";
    return 0;
}

Writing test_init_vectors.cpp


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[0, (n-1)n/2]
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]


Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed

    long long m = 0LL;                                                       //1

    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                //1+2n
        execints += 2;

        long long m = 0LL;

        execints++;
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n
            execints += 2;

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            execints++;

            execints++;
            if ( max_product < aux_product)  {                               //(n-1)n/2

				        max_product = aux_product;                                   //[0, (n-1)n/2]
                execints++;
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // --------------
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
}

int main() {                                                 //N inst exec by line executed
    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }

    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Writing max_pairwise_product_long_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
199999
100000
66667
50000
40000
33334
28572
25000
22223
20000
18182
2
2
2
2
2
2
2
2
1
0
80003672077


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077)

625028688.1015625

### Efficient solution

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile efficient_pw.cpp
#include <iostream>
#include <vector>
#include <fstream>
using namespace std;

int main () {
    int n;
    cin >> n;
    vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }
    int index = 1;
    for(int i =2; i<n; i++){
        if(numbers[i]> numbers[index]){
            index = i;
        }
        int temp = numbers[index];
        numbers[index] = numbers[n];
        numbers[n] = temp;
    }
    for(int i =2; i<(n-1); i++){
        if(numbers[i]> numbers[index]){
            index = i;
        }
        int temp = numbers[index];
        numbers[index] = numbers[n-1];
        numbers[n-1] = temp;
    }
    long long ans = (long long)numbers[n]*numbers[n-1];
    cout  << "The the maximum pairwise product is: "<< ans;
}

Writing efficient_pw.cpp


Test with inputs 1 and 2



In [ ]:
!g++ efficient_pw.cpp -o efficient_pw
!time ./efficient_pw < input2.txt
!g++ efficient_pw.cpp -o efficient_pw
!time ./efficient_pw < input3.txt

The the maximum pairwise product is: 11529800000
real	0m0.002s
user	0m0.002s
sys	0m0.000s
The the maximum pairwise product is: 39999800000
real	0m0.117s
user	0m0.099s
sys	0m0.005s


Test and time it with input 3

In [ ]:
!g++ efficient_pw.cpp -o efficient_pw
!time ./efficient_pw < input3.txt

The the maximum pairwise product is: 39999800000
real	0m0.075s
user	0m0.070s
sys	0m0.002s


20 stress tests

In [ ]:
%%writefile efficient_pw_test3.cpp

#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>
using namespace std;

int main () {
    srand(time(NULL));

    for (int i = 0; i < 20; ++i) {
        int n = rand() % 18 + 3; // Cantidad de elementos en la lista: 3 <= n <= 20
        vector<int> numbers(n);

        for (int j = 0; j < n; ++j) {
            numbers[j] = rand() % 1000;
        }

        int index = 1;
        for(int j = 2; j < n; j++){
            if(numbers[j] > numbers[index]){
                index = j;
            }
            int temp = numbers[index];
            numbers[index] = numbers[n - 1];
            numbers[n - 1] = temp;
        }

        for(int j = 2; j < n - 1; j++){
            if(numbers[j] > numbers[index]){
                index = j;
            }
            int temp = numbers[index];
            numbers[index] = numbers[n - 2];
            numbers[n - 2] = temp;
        }

        long long ans = (long long)numbers[n-1] * numbers[n-2];
        cout << "Caso " << i + 1 << ": ";
        for(int j = 0; j < n; j++){
            cout << numbers[j] << " ";
        }
        cout << "\nEl máximo producto de pares es: " << ans << "\n\n";
    }

    return 0;
}


Writing efficient_pw_test3.cpp


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
!g++ efficient_pw_test3.cpp -o efficient_pw_test3
!./efficient_pw_test3

Caso 1: 735 572 764 463 
El máximo producto de pares es: 353732

Caso 2: 10 159 304 36 89 342 596 255 478 38 789 477 740 717 804 441 
El máximo producto de pares es: 354564

Caso 3: 441 426 509 293 
El máximo producto de pares es: 149137

Caso 4: 537 815 576 306 820 148 422 718 571 432 877 913 903 
El máximo producto de pares es: 824439

Caso 5: 861 624 754 691 14 715 520 754 433 926 59 961 688 
El máximo producto de pares es: 661168

Caso 6: 549 48 47 86 215 392 123 118 623 814 
El máximo producto de pares es: 507122

Caso 7: 46 409 66 598 418 384 173 394 310 959 216 
El máximo producto de pares es: 207144

Caso 8: 737 515 811 
El máximo producto de pares es: 417665

Caso 9: 926 79 805 763 400 205 811 838 420 434 582 851 558 891 396 
El máximo producto de pares es: 352836

Caso 10: 346 145 764 882 916 
El máximo producto de pares es: 807912

Caso 11: 510 544 657 919 922 746 435 85 797 361 936 954 
El máximo producto de pares es: 892944

Caso 12: 688 159 288 526 579 722 
El máximo prod

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated number of instructions by second

1/(128/39999800000)

312498437.5

### Super efficient solution

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile super_efficient_pw.cpp
#include <iostream>
using namespace std;

int main() {
    int n;
    cin >> n; // 1 instrucción

    long long max1 = -1, max2 = -1, min1 = 1000001, min2 = 1000001; // 5 instrucciones

    for (int i = 0; i < n; ++i) { // n + 1 instrucciones
        long long num;
        cin >> num; // 1 instrucción

        if (num > max1) { // 1 instrucción
            max2 = max1; // 1 instrucción
            max1 = num; // 1 instrucción
        } else if (num > max2) { // 2 instrucciones
            max2 = num; // 1 instrucción
        }

        if (num < min1) { // 1 instrucción
            min2 = min1; // 1 instrucción
            min1 = num; // 1 instrucción
        } else if (num < min2) { // 2 instrucciones
            min2 = num; // 1 instrucción
        }
    }

    long long ans; // 1 instrucción
    if (max1 * max2 > min1 * min2) { // 4 instrucciones
        ans = max1 * max2; // 1 instrucción
    } else {
        ans = min1 * min2; // 1 instrucción
    }

    cout << ans << "\n"; // 2 instrucciones
    return 0; // 1 instrucción
}



Overwriting super_efficient_pw.cpp


Test with inputs 1 and 2




In [ ]:
!g++ super_efficient_pw.cpp -o super_efficient_pw
!time ./super_efficient_pw < input2.txt
!time ./super_efficient_pw < input3.txt

40000000000

real	0m0.002s
user	0m0.001s
sys	0m0.001s
39999800000

real	0m0.060s
user	0m0.056s
sys	0m0.002s


Test and time it with input 3

In [ ]:
!g++ super_efficient_pw.cpp -o super_efficient_pw
!time ./super_efficient_pw < input3.txt

39999800000

real	0m0.068s
user	0m0.066s
sys	0m0.002s


20 stress tests

In [ ]:
%%writefile super_efficient_pw_test3.cpp

#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>
using namespace std;

int main () {
    srand(time(NULL));

    for (int i = 0; i < 20; ++i) {
        int n = rand() % 18 + 3; // Cantidad de elementos en la lista: 3 <= n <= 20
        vector<int> numbers(n);

        for (int j = 0; j < n; ++j) {
            numbers[j] = rand() % 1000;
        }

        int index = 1;
        for(int j = 2; j < n; j++){
            if(numbers[j] > numbers[index]){
                index = j;
            }
            int temp = numbers[index];
            numbers[index] = numbers[n - 1];
            numbers[n - 1] = temp;
        }

        for(int j = 2; j < n - 1; j++){
            if(numbers[j] > numbers[index]){
                index = j;
            }
            int temp = numbers[index];
            numbers[index] = numbers[n - 2];
            numbers[n - 2] = temp;
        }

        long long ans = (long long)numbers[n-1] * numbers[n-2];
        cout << "Caso " << i + 1 << ": ";
        for(int j = 0; j < n; j++){
            cout << numbers[j] << " ";
        }
        cout << "\nEl máximo producto de pares es: " << ans << "\n\n";
    }

    return 0;
}

Overwriting super_efficient_pw_test3.cpp


In [ ]:
!g++ super_efficient_pw_test3.cpp -o super_efficient_pw_test3
!./super_efficient_pw_test3

Caso 1: 793 29 186 297 941 
El máximo producto de pares es: 279477

Caso 2: 23 689 516 699 140 620 182 560 865 104 213 500 592 618 985 804 
El máximo producto de pares es: 791940

Caso 3: 939 662 123 781 805 30 492 305 722 522 247 362 824 873 
El máximo producto de pares es: 719352

Caso 4: 211 531 19 352 503 221 264 369 426 477 788 970 
El máximo producto de pares es: 764360

Caso 5: 206 823 111 498 666 234 220 762 264 712 419 440 956 937 
El máximo producto de pares es: 895772

Caso 6: 459 329 539 
El máximo producto de pares es: 177331

Caso 7: 212 328 374 715 864 650 436 77 820 657 448 990 619 
El máximo producto de pares es: 612810

Caso 8: 730 362 580 316 
El máximo producto de pares es: 183280

Caso 9: 694 933 25 113 373 416 132 528 394 813 419 935 998 
El máximo producto de pares es: 933130

Caso 10: 791 740 179 133 750 529 
El máximo producto de pares es: 396750

Caso 11: 186 402 979 573 
El máximo producto de pares es: 560967

Caso 12: 655 764 534 972 50 580 905 48 46 983 278

Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile super_eficient_max_pairwise_product_long_execints.cpp
#include <iostream>
using namespace std;

int main() {
    int n;
    cin >> n; // 1 instrucción
    int contador = 1; // 1 instrucción

    long long max1 = -1, max2 = -1, min1 = 1000001, min2 = 1000001; // 5 instrucciones
    contador += 5; // 1 instrucción por cada variable creada

    for (int i = 0; i < n; ++i) { // n + 1 instrucciones
        long long num;
        cin >> num; // 1 instrucción
        contador += 2; // 1 instrucción por la lectura de la variable y la suma

        if (num > max1) { // 1 instrucción
            max2 = max1; // 1 instrucción
            max1 = num; // 1 instrucción
        } else if (num > max2) { // 2 instrucciones
            max2 = num; // 1 instrucción
        }

        if (num < min1) { // 1 instrucción
            min2 = min1; // 1 instrucción
            min1 = num; // 1 instrucción
        } else if (num < min2) { // 2 instrucciones
            min2 = num; // 1 instrucción
        }

        contador += 5; // 1 instrucción por cada operación realizada dentro del loop
    }

    long long ans; // 1 instrucción
    if (max1 * max2 > min1 * min2) { // 4 instrucciones
        ans = max1 * max2; // 1 instrucción
    } else {
        ans = min1 * min2; // 1 instrucción
    }

    cout << contador << "\n"; // 2 instrucciones
    return 0; // 1 instrucción
}

Writing super_eficient_max_pairwise_product_long_execints.cpp


In [ ]:
!g++ super_eficient_max_pairwise_product_long_execints.cpp -o super_eficient_max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
199999
100000
66667
50000
40000
33334
28572
25000
22223
20000
18182
2
2
2
2
2
2
2
2
1
0
80003672077


Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077

# Estimated number of instructions by second

1/(128/80003672077)

625028688.1015625